### PR TITLE
TLS verification of backend services

### PR DESCRIPTION
**Pull request by robbiemcmichael**
@davecheney It was good to meet you at the YOW! Conference. I briefly talked to you about implementing TLS verification for backend services; I finally got around to finishing off a sample implementation which verifies both the CA and hostname of the certificate.

This PR addresses issue #813. The implementation is similar to the design I suggested in that issue, however the ingress route schema has changed slightly to accommodate hostname verification. The question I had around the granularity of specifying the TLS verification settings (i.e. should it be one per ingress route, route or service?) is still my main concern and it would be good to get your thoughts. I'm happy to rework this PR based on any design concerns or code review comments.

### CODE REVIEW

**Comment by robbiemcmichael** [comment]
Updated my branch to remove some modifications to the deployment YAML file that were used for testing and accidentally left in as well as to add sign-offs to all commit messages.

File: internal/contour/cluster.go [hunk 1]
```diff
@@ -97,13 +97,7 @@ func visitClusters(root dag.Vertex) map[string]*v2.Cluster {
...
 func (v *clusterVisitor) visit(vertex dag.Vertex) {
 	switch service := vertex.(type) {
-	case *dag.HTTPService:
-		name := envoy.Clustername(&service.TCPService)
-		if _, ok := v.clusters[name]; !ok {
-			c := envoy.Cluster(service)
-			v.clusters[c.Name] = c
-		}
-	case *dag.TCPService:
+	case dag.Service:
```
Comment (davecheney) [inline internal/contour/cluster.go]: This is good cleanup, but unrelated to this change. Can you please send it as its own PR. Thank you.

Comment (robbiemcmichael) [reply internal/contour/cluster.go]: It wasn't completely unrelated to this PR, I had to change `Clustername` to take a `Service` as a parameter since the hash needed to include [fields that are only present in an `HTTPService`](https://github.com/heptio/contour/pull/852/commits/b7be44985e81059703fd5c7f003a872c820d3b7a#diff-0f385e580f0bf79e437d3ad4d0b52f20R134).

The cleanup here was just a nice byproduct of changing that function, but I can still split that part out into a separate PR if you'd like.

**Comment by davecheney** [comment]
Thank you for working on this. I'm sorry that it took so long to resolve and that we ended up duplicating some of your work.

I'm going to close this as a duplicate of #1045
